Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/diagnosiskeyprovider/DefaultDiagnosisKeyProvider.kt

```diff
@@ -22,95 +19,37 @@ class DefaultDiagnosisKeyProvider @Inject constructor(
     private val enfClient: ExposureNotificationClient
 ) : DiagnosisKeyProvider {
 
-    override suspend fun provideDiagnosisKeys(
-        keyFiles: Collection<File>,
-        configuration: ExposureConfiguration?,
-        token: String
-    ): Boolean {
-        return try {
-            if (keyFiles.isEmpty()) {
-                Timber.d("No key files submitted, returning early.")
-                return true
-            }
-
-            val usedConfiguration = if (configuration == null) {
-                Timber.w("Passed configuration was NULL, creating fallback.")
-                ExposureConfiguration.ExposureConfigurationBuilder().build()
-            } else {
-                configuration
-            }
-
-            if (googleAPIVersion.isAtLeast(GoogleAPIVersion.V16)) {
-                provideKeys(keyFiles, usedConfiguration, token)
-            } else {
-                provideKeysLegacy(keyFiles, usedConfiguration, token)
-            }
-        } catch (e: Exception) {
-            Timber.e(
-                e, "Error during provideDiagnosisKeys(keyFiles=%s, configuration=%s, token=%s)",
-                keyFiles, configuration, token
-            )
-            throw e
+    override suspend fun provideDiagnosisKeys(keyFiles: Collection<File>): Boolean {
+        if (keyFiles.isEmpty()) {
+            Timber.d("No key files submitted, returning early.")
+            return true
         }
-    }
 
-    private suspend fun provideKeys(
-        files: Collection<File>,
-        configuration: ExposureConfiguration,
-        token: String
-    ): Boolean {
-        Timber.d("Using non-legacy key provision.")
-
-        if (!submissionQuota.consumeQuota(1)) {
-            Timber.w("Not enough quota available.")
-            // TODO Currently only logging, we'll be more strict in a future release
-            // return false
+        if (!googleAPIVersion.isAtLeast(GoogleAPIVersion.V15)) {
+            // Actually this shouldn't happen
+            Timber.d("No key files submitted because client uses an old unsupported version")
+            return false
```

Comment:
I think we should throw an exception here, this is worthy of making it user visible as the app will not be able to function. 

We could create `googleAPIVersion.requireV15()` that throws a specific exception if the requirement is not satisfied. :thinking: 

What do you think?

Reply:
Is it ok to leave my 2 cents here?

l29: `// Actually this shouldn't happen`
This could happen when users are reactivating old phones to provide CWA to family members without phone.

Why not lifting UI/UX to the next level, and present a nice 'information (error) card' in the home fragment for known problems like API 17 or API 39508 instead of throwing an exception? Much work, yes, but as #1615 is WIP, probably a good point in time to think about it?

Reply:
I'm also not sure about this.. the check feels like a second or third check in dept.. even with an exception the user would be left a little alone here. Sure, better than crashing, but imho it makes more sense to check all the API/enf requirements while the app is starting, what do y think?

Reply:
AFAICT (please confirm), if we throw here, we crash the task, and the crashed task will be caught by `TaskController` which will call `Error.report...` which will cause the "popular" error dialog to show, it will not crash the app. So my favored approach would be to just throw an exception here with a specific error message such that it's clear to the user what happened. Something like "Your systems exposure notification service are too old". This would be an easy fix and enough to let users reactivating old devices what is happening, which is not impossible, but probably rare.

I don't think this warrants more elaborate UI elements. Although I don't like our error dialog, and @BMItter has already been working on a nicer solution, it's not ready yet due to shifting priorities.

Thoughts @BMItter @vaubaehn ?

Reply:
please continue discussion at https://github.com/corona-warn-app/cwa-app-android/pull/1659/files#r526719742